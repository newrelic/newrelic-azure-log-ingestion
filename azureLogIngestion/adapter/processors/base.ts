import { Context } from "@azure/functions"

export interface Processor {
    processMessage(message: Record<string, any>, context: Context): void
    sendBatch(): Promise<void>
}
