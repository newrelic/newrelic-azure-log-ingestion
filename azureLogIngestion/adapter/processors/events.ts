import { telemetry } from "@newrelic/telemetry-sdk"
import { Context } from "@azure/functions"

import { Processor } from "./base"
import flatten from "../../utils/flatten"

export default class EventProcessor implements Processor {
    client: telemetry.events.EventClient
    batch: telemetry.events.EventBatch

    constructor(apiKey: string) {
        const host =
            process.env.NEW_RELIC_REGION === "eu"
                ? "insights-collector.eu01.nr-data.net"
                : "insights-collector.nr-data.net"
        this.client = new telemetry.events.EventClient({ apiKey, host })
        this.startNewBatch()
    }

    /**
     * Starts a new batch, setting common attributes
     */
    private startNewBatch(): void {
        this.batch = new telemetry.events.EventBatch({ "cloud.provider": "azure" })
    }

    /**
     * Processes an event message and adds event to current batch
     */
    processMessage(message: Record<string, any>, context: Context): void {
        // Deleting attributes we do not want to send to New Relic
        // TODO: Make this a part of a processor attribute filter method
        delete message.iKey

        const { type, timestamp, properties = {}, ...rest } = message
        const eventType = `Azure${type}`
        const epochDate = new Date(timestamp).getTime()
        const attributes = {
            ...flatten({ ...properties, ...rest }),
        }

        const event = new telemetry.events.Event(eventType, attributes, epochDate)
        this.batch.addEvent(event)
    }

    /**
     * Sends event telemetry batches to New Relic
     *
     * Currently this doesn't create a new batch on failure, allowing for
     * reattempts. This will result in a memory leak for subsequent failures.
     * Should set a max attempt count and start new batch when reached.
     */
    sendBatch(): Promise<void> {
        return new Promise<void>((resolve, reject) => {
            this.client.send(this.batch, (err) => {
                if (err) reject(err)
            })

            this.startNewBatch()
            resolve()
        })
    }
}
