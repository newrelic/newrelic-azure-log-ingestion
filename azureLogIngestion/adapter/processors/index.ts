export { default as EventProcessor } from "./events"
export { default as SpanProcessor } from "./spans"
