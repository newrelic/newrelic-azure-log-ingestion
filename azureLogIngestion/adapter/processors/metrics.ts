import { telemetry } from "@newrelic/telemetry-sdk"
import { Context } from "@azure/functions"
import { Processor } from "./base"
import flatten from "../../utils/flatten"
import { GaugeMetric, CountMetric, SummaryMetric } from "@newrelic/telemetry-sdk/dist/src/telemetry/metrics"

import { convertToMs } from "../../utils/time"

export default class MetricsProcessor implements Processor {
    client: telemetry.metrics.MetricClient
    batch: telemetry.metrics.MetricBatch

    constructor(apiKey: string) {
        const host = process.env.NEW_RELIC_REGION === "eu" ? "metric-api.eu.newrelic.com" : "metric-api.newrelic.com"
        this.client = new telemetry.metrics.MetricClient({ apiKey, host })
        this.startNewBatch()
    }

    /**
     * Starts a new batch, setting common attributes
     */
    private startNewBatch(): void {
        this.batch = new telemetry.metrics.MetricBatch({ "cloud.provider": "azure" })
    }

    private performanceCounter(message: any): GaugeMetric | CountMetric {
        const { name, value, type, category, timestamp, properties, ...rest } = message
        const epochDate = new Date(timestamp).getTime()
        const attributes = {
            ...flatten({ name, value, type, category, ...properties, ...rest, timestamp: epochDate }),
        }
        if ((rest && rest.interval) || (properties && properties.interval)) {
            const intVal = (rest && rest.interval) || (properties && properties.interval)
            const intervalMs = convertToMs(intVal)
            return new telemetry.metrics.CountMetric(name, value, attributes, epochDate, intervalMs)
        }
        // Most performance counters are likely to be gauges.
        return new telemetry.metrics.GaugeMetric(name, value, attributes, epochDate)
    }

    private metric(message: any): GaugeMetric | CountMetric | SummaryMetric {
        const { name, value, min, max, sum, itemCount, interval, timestamp, properties, ...rest } = message
        let { count } = message
        let intervalMs
        // example AppMetrics we've seen use itemCount instead of count
        if (!count) {
            count = itemCount
        }
        const epochDate = new Date(timestamp).getTime()
        if (interval) {
            intervalMs = convertToMs(interval)
        } else {
            // otherwise, assume default 10s.
            intervalMs = 10000
        }
        const attributes = {
            ...flatten({ name, ...properties, ...rest, timestamp: epochDate, intervalMs }),
        }
        if (value !== undefined) attributes.value = value
        if (sum !== undefined) attributes.sum = sum
        if (min !== undefined) attributes.min = min
        if (max !== undefined) attributes.max = max

        // this could be a summary metric
        if (count || sum || min || max) {
            return new telemetry.metrics.SummaryMetric(
                name,
                { count, sum, min, max },
                attributes,
                epochDate,
                intervalMs,
            )
        }
        if (value) {
            return new telemetry.metrics.CountMetric(name, value, attributes, epochDate, intervalMs)
        }
    }

    private createMetric(message: any): GaugeMetric | CountMetric | SummaryMetric {
        const { type } = message
        if (["AppPerformanceCounter", "appPerformanceCounter"].includes(type)) {
            return this.performanceCounter(message)
        }
        return this.metric(message)
    }

    /**
     * Processes a metric and add to current batch
     */
    processMessage(message: Record<string, any>, context: Context): void {
        // Deleting attributes we do not want to send to New Relic
        // TODO: Make this a part of a processor attribute filter method
        delete message.iKey
        const metric = this.createMetric(message)
        this.batch.addMetric(metric)
    }

    /**
     * Sends event telemetry batches to New Relic
     *
     * Currently this doesn't create a new batch on failure, allowing for
     * reattempts. This will result in a memory leak for subsequent failures.
     * Should set a max attempt count and start new batch when reached.
     */
    sendBatch(): Promise<void> {
        return new Promise<void>((resolve, reject) => {
            this.client.send(this.batch, (err) => {
                if (err) reject(err)
            })
            this.startNewBatch()
            resolve()
        })
    }
}
