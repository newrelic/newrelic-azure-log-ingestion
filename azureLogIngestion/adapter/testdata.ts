export const appRequest = `{
  "records": [
    {
      "time": "2021-04-29T22:41:14.1238381Z",
      "resourceId": "/SUBSCRIPTIONS/subscriptionID goes here/RESOURCEGROUPS/resource group goes here/PROVIDERS/MICROSOFT.INSIGHTS/COMPONENTS/{function app name}",
      "ResourceGUID": "resourceGuid",
      "Type": "AppRequests",
      "AppRoleInstance": "very long id string",
      "AppRoleName": "functionAppName goes here",
      "ClientCity": "Portland",
      "ClientCountryOrRegion": "United States",
      "ClientIP": "0.0.0.0",
      "ClientStateOrProvince": "Oregon",
      "ClientType": "PC",
      "IKey": "iKeyValue",
      "_BilledSize": 1082,
      "OperationName": "{function name}",
      "OperationId": "operationIdValue",
      "ParentId": "operationIdValue",
      "SDKVersion": "azurefunctions: 3.0.15571.0",
      "Properties": {
        "LogLevel": "Information",
        "HttpPath": "/api/{function name goes here}",
        "FunctionExecutionTimeMs": "4.7239",
        "TriggerReason": "This function was programmatically called via the host APIs.",
        "InvocationId": "invocationIdValue",
        "ProcessId": "8912",
        "Category": "Host.Results",
        "HttpMethod": "GET",
        "FullName": "Functions.{function name}",
        "HostInstanceId": "hostInstanceIdValue"
      },
      "Id": "09d760d34d75e847",
      "Name": "{function name}",
      "Url": "https://{function app name}.azurewebsites.net/api/{function name}",
      "Success": true,
      "ResultCode": "200",
      "DurationMs": 9.1593,
      "PerformanceBucket": "<250ms",
      "ItemCount": 1
    }
  ]
}`
