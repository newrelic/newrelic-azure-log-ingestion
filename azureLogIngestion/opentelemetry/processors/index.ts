export { default as SpanProcessor } from "./spans"
