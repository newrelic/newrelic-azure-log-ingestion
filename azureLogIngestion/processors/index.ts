export { default as EventProcessor } from "./events"
export { default as SpanProcessor } from "./spans"
export { default as LogProcessor } from "./logs"
export { default as MetricsProcessor } from "./metrics"
