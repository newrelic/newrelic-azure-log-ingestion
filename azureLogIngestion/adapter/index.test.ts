import { telemetry } from "@newrelic/telemetry-sdk"

import Adapter from "./index"
import { EventProcessor, SpanProcessor } from "./processors"

import { appInsightsAppDependency, appInsightsAppEvent, appInsightsAppRequest, arrayOfStrings } from "./testdata.test"

describe("Adapter", () => {
    it("instantiates with api key", () => {
        const adapter = new Adapter("mock-insert-key")

        expect(adapter).toBeInstanceOf(Adapter)

        expect(adapter.eventProcessor).toBeInstanceOf(EventProcessor)
        expect(adapter.eventProcessor.batch).toBeInstanceOf(telemetry.events.EventBatch)
        expect(adapter.eventProcessor.batch.getBatchSize()).toEqual(0)
        expect(adapter.eventProcessor.client).toBeInstanceOf(telemetry.events.EventClient)

        expect(adapter.spanProcessor).toBeInstanceOf(SpanProcessor)
        expect(adapter.spanProcessor.batch).toBeInstanceOf(telemetry.spans.SpanBatch)
        expect(adapter.spanProcessor.batch.getBatchSize()).toEqual(0)
        expect(adapter.spanProcessor.client).toBeInstanceOf(telemetry.spans.SpanClient)
    })

    it("processes app request and dependency as spans", () => {
        const adapter = new Adapter("mock-insert-key")

        const log = (...args) => null
        log.verbose = (...args) => null
        log.info = (...args) => null
        log.warn = (...args) => null
        log.error = (...args) => null

        const mockContext = {
            bindings: {},
            bindingData: {},
            bindingDefinitions: [],
            done: () => null,
            executionContext: { invocationId: "foobar", functionName: "foobar", functionDirectory: "foobar" },
            invocationId: "foobar",
            log,
            traceContext: { attributes: {}, traceparent: "foobar", tracestate: "foobar" },
        }

        expect(adapter.spanProcessor.batch.getBatchSize()).toEqual(0)
        expect(adapter.spanProcessor.batch.spans).toMatchSnapshot()

        adapter.processMessages(appInsightsAppRequest, mockContext)
        expect(adapter.spanProcessor.batch.getBatchSize()).toEqual(1)
        expect(adapter.spanProcessor.batch.spans).toMatchSnapshot()

        adapter.processMessages(appInsightsAppDependency, mockContext)
        expect(adapter.spanProcessor.batch.getBatchSize()).toEqual(2)
        expect(adapter.spanProcessor.batch.spans).toMatchSnapshot()
    })

    it("processes app request as event", () => {
        const adapter = new Adapter("mock-insert-key")

        const log = (...args) => null
        log.verbose = (...args) => null
        log.info = (...args) => null
        log.warn = (...args) => null
        log.error = (...args) => null

        const mockContext = {
            bindings: {},
            bindingData: {},
            bindingDefinitions: [],
            done: () => null,
            executionContext: { invocationId: "foobar", functionName: "foobar", functionDirectory: "foobar" },
            invocationId: "foobar",
            log,
            traceContext: { attributes: {}, traceparent: "foobar", tracestate: "foobar" },
        }

        expect(adapter.eventProcessor.batch.getBatchSize()).toEqual(0)
        expect(adapter.eventProcessor.batch.events).toMatchSnapshot()

        adapter.processMessages(appInsightsAppRequest, mockContext)
        expect(adapter.eventProcessor.batch.getBatchSize()).toEqual(1)
        expect(adapter.eventProcessor.batch.events).toMatchSnapshot()
    })

    it("processes app request and dependency when sent as array of strings", () => {
        const adapter = new Adapter("mock-insert-key")

        const log = (...args) => null
        log.verbose = (...args) => null
        log.info = (...args) => null
        log.warn = (...args) => null
        log.error = (...args) => null

        const mockContext = {
            bindings: {},
            bindingData: {},
            bindingDefinitions: [],
            done: () => null,
            executionContext: { invocationId: "foobar", functionName: "foobar", functionDirectory: "foobar" },
            invocationId: "foobar",
            log,
            traceContext: { attributes: {}, traceparent: "foobar", tracestate: "foobar" },
        }

        expect(adapter.spanProcessor.batch.getBatchSize()).toEqual(0)
        expect(adapter.spanProcessor.batch.spans).toMatchSnapshot()

        adapter.processMessages(arrayOfStrings, mockContext)
        expect(adapter.spanProcessor.batch.getBatchSize()).toEqual(3)
        expect(adapter.spanProcessor.batch.spans).toMatchSnapshot()

        adapter.processMessages(appInsightsAppDependency, mockContext)
        expect(adapter.spanProcessor.batch.getBatchSize()).toEqual(4)
        expect(adapter.spanProcessor.batch.spans).toMatchSnapshot()
    })

    it("processes app event as event and span", () => {
        const adapter = new Adapter("mock-insert-key")

        const log = (...args) => null
        log.verbose = (...args) => null
        log.info = (...args) => null
        log.warn = (...args) => null
        log.error = (...args) => null

        const mockContext = {
            bindings: {},
            bindingData: {},
            bindingDefinitions: [],
            done: () => null,
            executionContext: { invocationId: "foobar", functionName: "foobar", functionDirectory: "foobar" },
            invocationId: "foobar",
            log,
            traceContext: { attributes: {}, traceparent: "foobar", tracestate: "foobar" },
        }

        expect(adapter.eventProcessor.batch.getBatchSize()).toEqual(0)
        expect(adapter.spanProcessor.batch.getBatchSize()).toEqual(0)
        expect(adapter.eventProcessor.batch.events).toMatchSnapshot()
        expect(adapter.spanProcessor.batch.spans).toMatchSnapshot()

        adapter.processMessages(appInsightsAppEvent, mockContext)

        expect(adapter.eventProcessor.batch.getBatchSize()).toEqual(1)
        expect(adapter.spanProcessor.batch.getBatchSize()).toEqual(1)
        expect(adapter.eventProcessor.batch.events).toMatchSnapshot()
        expect(adapter.spanProcessor.batch.spans).toMatchSnapshot()
    })
})
