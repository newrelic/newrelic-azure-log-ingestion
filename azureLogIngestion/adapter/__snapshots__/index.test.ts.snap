// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Adapter processes app event as event and span 1`] = `Array []`;

exports[`Adapter processes app event as event and span 2`] = `Array []`;

exports[`Adapter processes app event as event and span 3`] = `
Array [
  Event {
    "attributes": Object {
      "appRoleInstance": "{role instance value}",
      "appRoleName": "Web",
      "billedSize": 742,
      "clientCity": "San Jose",
      "clientCountryOrRegion": "United States",
      "clientIp": "0.0.0.0",
      "clientOs": "Windows_NT 10.0.14393",
      "clientStateOrProvince": "California",
      "clientType": "PC",
      "itemCount": 1,
      "myCustomProperty1": "This is my custom property",
      "myCustomProperty2": "And here is another.",
      "name": "my custom event",
      "now": "2021-05-04T19:12:36.185Z",
      "operationId": "98765",
      "parentId": "12345",
      "resourceGuid": "{guid of resource}",
      "resourceId": "/SUBSCRIPTIONS/{subscription Id}/RESOURCEGROUPS/{resource group in all caps}/PROVIDERS/MICROSOFT.INSIGHTS/COMPONENTS/{function app name in all caps}",
      "sdkVersion": "node:1.8.10",
    },
    "eventType": "AzureAppEvent",
    "timestamp": 1620155556185,
  },
]
`;

exports[`Adapter processes app event as event and span 4`] = `
Array [
  Span {
    "attributes": Object {
      "appRoleInstance": "{role instance value}",
      "appRoleName": "Web",
      "billedSize": 742,
      "clientCity": "San Jose",
      "clientCountryOrRegion": "United States",
      "clientIp": "0.0.0.0",
      "clientOs": "Windows_NT 10.0.14393",
      "clientStateOrProvince": "California",
      "clientType": "PC",
      "itemCount": 1,
      "name": "my custom event",
      "parent.id": "12345",
      "properties.myCustomProperty1": "This is my custom property",
      "properties.myCustomProperty2": "And here is another.",
      "properties.now": "2021-05-04T19:12:36.185Z",
      "resourceGuid": "{guid of resource}",
      "resourceId": "/SUBSCRIPTIONS/{subscription Id}/RESOURCEGROUPS/{resource group in all caps}/PROVIDERS/MICROSOFT.INSIGHTS/COMPONENTS/{function app name in all caps}",
      "sdkVersion": "node:1.8.10",
      "type": "AppEvent",
    },
    "id": undefined,
    "timestamp": 1620155556185,
    "trace.id": "98765",
  },
]
`;

exports[`Adapter processes app request and dependency as spans 1`] = `Array []`;

exports[`Adapter processes app request and dependency as spans 2`] = `
Array [
  Span {
    "attributes": Object {
      "appRoleInstance": "very long id string",
      "appRoleName": "functionAppName goes here",
      "billedSize": 1082,
      "clientCity": "Portland",
      "clientCountryOrRegion": "United States",
      "clientIp": "0.0.0.0",
      "clientStateOrProvince": "Oregon",
      "clientType": "PC",
      "duration.ms": 9.1593,
      "http.statusCode": "200",
      "http.url": "https://{function app name}.azurewebsites.net/api/{function name}",
      "itemCount": 1,
      "name": "{function name}",
      "performanceBucket": "<250ms",
      "properties.category": "Host.Results",
      "properties.fullName": "Functions.{function name}",
      "properties.functionExecutionTimeMs": "4.7239",
      "properties.hostInstanceId": "hostInstanceIdValue",
      "properties.http.method": "GET",
      "properties.http.path": "/api/{function name goes here}",
      "properties.invocationId": "invocationIdValue",
      "properties.logLevel": "Information",
      "properties.processId": "8912",
      "properties.triggerReason": "This function was programmatically called via the host APIs.",
      "resourceGuid": "resourceGuid",
      "resourceId": "/SUBSCRIPTIONS/subscriptionID goes here/RESOURCEGROUPS/resource group goes here/PROVIDERS/MICROSOFT.INSIGHTS/COMPONENTS/{function app name}",
      "sdkVersion": "azurefunctions: 3.0.15571.0",
      "service.name": "{function name}",
      "success": true,
      "type": "AppRequest",
    },
    "id": "09d760d34d75e847",
    "timestamp": 1619736074123,
    "trace.id": "12345",
  },
]
`;

exports[`Adapter processes app request and dependency as spans 3`] = `
Array [
  Span {
    "attributes": Object {
      "appRoleInstance": "very long id string",
      "appRoleName": "functionAppName goes here",
      "billedSize": 1082,
      "clientCity": "Portland",
      "clientCountryOrRegion": "United States",
      "clientIp": "0.0.0.0",
      "clientStateOrProvince": "Oregon",
      "clientType": "PC",
      "duration.ms": 9.1593,
      "http.statusCode": "200",
      "http.url": "https://{function app name}.azurewebsites.net/api/{function name}",
      "itemCount": 1,
      "name": "{function name}",
      "performanceBucket": "<250ms",
      "properties.category": "Host.Results",
      "properties.fullName": "Functions.{function name}",
      "properties.functionExecutionTimeMs": "4.7239",
      "properties.hostInstanceId": "hostInstanceIdValue",
      "properties.http.method": "GET",
      "properties.http.path": "/api/{function name goes here}",
      "properties.invocationId": "invocationIdValue",
      "properties.logLevel": "Information",
      "properties.processId": "8912",
      "properties.triggerReason": "This function was programmatically called via the host APIs.",
      "resourceGuid": "resourceGuid",
      "resourceId": "/SUBSCRIPTIONS/subscriptionID goes here/RESOURCEGROUPS/resource group goes here/PROVIDERS/MICROSOFT.INSIGHTS/COMPONENTS/{function app name}",
      "sdkVersion": "azurefunctions: 3.0.15571.0",
      "service.name": "{function name}",
      "success": true,
      "type": "AppRequest",
    },
    "id": "09d760d34d75e847",
    "timestamp": 1619736074123,
    "trace.id": "12345",
  },
  Span {
    "attributes": Object {
      "appRoleInstance": "{role instance id}",
      "appRoleName": "Web",
      "billedSize": 786,
      "clientCity": "San Jose",
      "clientCountryOrRegion": "United States",
      "clientIp": "0.0.0.0",
      "clientOs": "Windows_NT 10.0.14393",
      "clientStateOrProvince": "California",
      "clientType": "PC",
      "dependencyType": "http",
      "duration.ms": 231,
      "http.target": "{target property of client.trackDependency}",
      "itemCount": 1,
      "name": "External/{name property of client.trackDependency}",
      "parent.id": "12345",
      "performanceBucket": "<250ms",
      "resourceGuid": "{resource uuid4}",
      "resourceId": "/SUBSCRIPTIONS/{subscription id}/RESOURCEGROUPS/{resource group in all caps}/PROVIDERS/MICROSOFT.INSIGHTS/COMPONENTS/{function app name in all caps}",
      "sdkVersion": "node:1.8.10",
      "success": true,
      "type": "AppDependency",
    },
    "id": "{id value}",
    "timestamp": 1620165670940,
    "trace.id": "67890",
  },
]
`;

exports[`Adapter processes app request and dependency when sent as array of strings 1`] = `Array []`;

exports[`Adapter processes app request and dependency when sent as array of strings 2`] = `
Array [
  Span {
    "attributes": Object {
      "appRoleInstance": "roleInstanceId",
      "appRoleName": "Web",
      "billedSize": 786,
      "clientCity": "San Jose",
      "clientCountryOrRegion": "United States",
      "clientIp": "0.0.0.0",
      "clientOs": "Windows_NT 10.0.14393",
      "clientStateOrProvince": "California",
      "clientType": "PC",
      "dependencyType": "Buffer",
      "duration.ms": 231,
      "itemCount": 1,
      "name": "node buffer",
      "performanceBucket": "<250ms",
      "resourceGuid": "resourceGuid",
      "resourceId": "/SUBSCRIPTIONS/subscriptionId/RESOURCEGROUPS/resourceGroupName/PROVIDERS/MICROSOFT.INSIGHTS/COMPONENTS/functionAppName",
      "sdkVersion": "node:1.8.10",
      "success": true,
      "target": "Buffer",
      "type": "AppDependency",
    },
    "id": "recordId",
    "timestamp": 1620674371908,
    "trace.id": "parentId",
  },
  Span {
    "attributes": Object {
      "appRoleInstance": "dw0-hr0-1068-25",
      "appRoleName": "Web",
      "billedSize": 834,
      "clientCity": "San Jose",
      "clientCountryOrRegion": "United States",
      "clientIp": "0.0.0.0",
      "clientOs": "Windows_NT 10.0.14393",
      "clientStateOrProvince": "California",
      "clientType": "PC",
      "data": "SELECT * FROM Customers",
      "dependencyType": "ZSQL",
      "duration.ms": 231,
      "itemCount": 1,
      "name": "select customers proc",
      "parent.id": "parentId",
      "performanceBucket": "<250ms",
      "resourceGuid": "resourceGuid",
      "resourceId": "/SUBSCRIPTIONS/subscriptionId/RESOURCEGROUPS/resourceGroupName/PROVIDERS/MICROSOFT.INSIGHTS/COMPONENTS/functionAppName",
      "sdkVersion": "node:1.8.10",
      "success": true,
      "target": "http://dbname",
      "type": "AppDependency",
    },
    "id": "id",
    "timestamp": 1620674372049,
    "trace.id": "opId",
  },
  Span {
    "attributes": Object {
      "appRoleInstance": "roleInstanceId",
      "appRoleName": "functionAppName",
      "billedSize": 1087,
      "clientCity": "Pittsburgh",
      "clientCountryOrRegion": "United States",
      "clientIp": "0.0.0.0",
      "clientStateOrProvince": "Pennsylvania",
      "clientType": "PC",
      "duration.ms": 30021.5457,
      "http.statusCode": "200",
      "http.url": "https://functionAppName.azurewebsites.net/api/functionName",
      "itemCount": 1,
      "name": "functionName",
      "parent.id": "parentId",
      "performanceBucket": "30sec-1min",
      "properties.category": "Host.Results",
      "properties.fullName": "Functions.functionName",
      "properties.functionExecutionTimeMs": "30020.222",
      "properties.hostInstanceId": "hostInstance",
      "properties.http.method": "GET",
      "properties.http.path": "/api/functionName",
      "properties.invocationId": "invocationId",
      "properties.logLevel": "Information",
      "properties.processId": "6904",
      "properties.triggerReason": "This function was programmatically called via the host APIs.",
      "resourceGuid": "resourceGuid",
      "resourceId": "/SUBSCRIPTIONS/subscriptionId/RESOURCEGROUPS/resourceGroupName/PROVIDERS/MICROSOFT.INSIGHTS/COMPONENTS/functionAppName",
      "sdkVersion": "azurefunctions: 3.0.15584.0",
      "service.name": "functionName",
      "success": true,
      "type": "AppRequest",
    },
    "id": "bc40802d79559544",
    "timestamp": 1620674361060,
    "trace.id": "operationId",
  },
]
`;

exports[`Adapter processes app request and dependency when sent as array of strings 3`] = `
Array [
  Span {
    "attributes": Object {
      "appRoleInstance": "roleInstanceId",
      "appRoleName": "Web",
      "billedSize": 786,
      "clientCity": "San Jose",
      "clientCountryOrRegion": "United States",
      "clientIp": "0.0.0.0",
      "clientOs": "Windows_NT 10.0.14393",
      "clientStateOrProvince": "California",
      "clientType": "PC",
      "dependencyType": "Buffer",
      "duration.ms": 231,
      "itemCount": 1,
      "name": "node buffer",
      "performanceBucket": "<250ms",
      "resourceGuid": "resourceGuid",
      "resourceId": "/SUBSCRIPTIONS/subscriptionId/RESOURCEGROUPS/resourceGroupName/PROVIDERS/MICROSOFT.INSIGHTS/COMPONENTS/functionAppName",
      "sdkVersion": "node:1.8.10",
      "success": true,
      "target": "Buffer",
      "type": "AppDependency",
    },
    "id": "recordId",
    "timestamp": 1620674371908,
    "trace.id": "parentId",
  },
  Span {
    "attributes": Object {
      "appRoleInstance": "dw0-hr0-1068-25",
      "appRoleName": "Web",
      "billedSize": 834,
      "clientCity": "San Jose",
      "clientCountryOrRegion": "United States",
      "clientIp": "0.0.0.0",
      "clientOs": "Windows_NT 10.0.14393",
      "clientStateOrProvince": "California",
      "clientType": "PC",
      "data": "SELECT * FROM Customers",
      "dependencyType": "ZSQL",
      "duration.ms": 231,
      "itemCount": 1,
      "name": "select customers proc",
      "parent.id": "parentId",
      "performanceBucket": "<250ms",
      "resourceGuid": "resourceGuid",
      "resourceId": "/SUBSCRIPTIONS/subscriptionId/RESOURCEGROUPS/resourceGroupName/PROVIDERS/MICROSOFT.INSIGHTS/COMPONENTS/functionAppName",
      "sdkVersion": "node:1.8.10",
      "success": true,
      "target": "http://dbname",
      "type": "AppDependency",
    },
    "id": "id",
    "timestamp": 1620674372049,
    "trace.id": "opId",
  },
  Span {
    "attributes": Object {
      "appRoleInstance": "roleInstanceId",
      "appRoleName": "functionAppName",
      "billedSize": 1087,
      "clientCity": "Pittsburgh",
      "clientCountryOrRegion": "United States",
      "clientIp": "0.0.0.0",
      "clientStateOrProvince": "Pennsylvania",
      "clientType": "PC",
      "duration.ms": 30021.5457,
      "http.statusCode": "200",
      "http.url": "https://functionAppName.azurewebsites.net/api/functionName",
      "itemCount": 1,
      "name": "functionName",
      "parent.id": "parentId",
      "performanceBucket": "30sec-1min",
      "properties.category": "Host.Results",
      "properties.fullName": "Functions.functionName",
      "properties.functionExecutionTimeMs": "30020.222",
      "properties.hostInstanceId": "hostInstance",
      "properties.http.method": "GET",
      "properties.http.path": "/api/functionName",
      "properties.invocationId": "invocationId",
      "properties.logLevel": "Information",
      "properties.processId": "6904",
      "properties.triggerReason": "This function was programmatically called via the host APIs.",
      "resourceGuid": "resourceGuid",
      "resourceId": "/SUBSCRIPTIONS/subscriptionId/RESOURCEGROUPS/resourceGroupName/PROVIDERS/MICROSOFT.INSIGHTS/COMPONENTS/functionAppName",
      "sdkVersion": "azurefunctions: 3.0.15584.0",
      "service.name": "functionName",
      "success": true,
      "type": "AppRequest",
    },
    "id": "bc40802d79559544",
    "timestamp": 1620674361060,
    "trace.id": "operationId",
  },
  Span {
    "attributes": Object {
      "appRoleInstance": "{role instance id}",
      "appRoleName": "Web",
      "billedSize": 786,
      "clientCity": "San Jose",
      "clientCountryOrRegion": "United States",
      "clientIp": "0.0.0.0",
      "clientOs": "Windows_NT 10.0.14393",
      "clientStateOrProvince": "California",
      "clientType": "PC",
      "dependencyType": "http",
      "duration.ms": 231,
      "http.target": "{target property of client.trackDependency}",
      "itemCount": 1,
      "name": "External/{name property of client.trackDependency}",
      "parent.id": "12345",
      "performanceBucket": "<250ms",
      "resourceGuid": "{resource uuid4}",
      "resourceId": "/SUBSCRIPTIONS/{subscription id}/RESOURCEGROUPS/{resource group in all caps}/PROVIDERS/MICROSOFT.INSIGHTS/COMPONENTS/{function app name in all caps}",
      "sdkVersion": "node:1.8.10",
      "success": true,
      "type": "AppDependency",
    },
    "id": "{id value}",
    "timestamp": 1620165670940,
    "trace.id": "67890",
  },
]
`;

exports[`Adapter processes app request as event 1`] = `Array []`;

exports[`Adapter processes app request as event 2`] = `
Array [
  Event {
    "attributes": Object {
      "appRoleInstance": "very long id string",
      "appRoleName": "functionAppName goes here",
      "billedSize": 1082,
      "category": "Host.Results",
      "clientCity": "Portland",
      "clientCountryOrRegion": "United States",
      "clientIp": "0.0.0.0",
      "clientStateOrProvince": "Oregon",
      "clientType": "PC",
      "durationMs": 9.1593,
      "fullName": "Functions.{function name}",
      "functionExecutionTimeMs": "4.7239",
      "hostInstanceId": "hostInstanceIdValue",
      "http.method": "GET",
      "http.path": "/api/{function name goes here}",
      "http.statusCode": "200",
      "http.url": "https://{function app name}.azurewebsites.net/api/{function name}",
      "id": "09d760d34d75e847",
      "invocationId": "invocationIdValue",
      "itemCount": 1,
      "logLevel": "Information",
      "name": "{function name}",
      "operationId": "12345",
      "operationName": "{function name}",
      "parentId": "12345",
      "performanceBucket": "<250ms",
      "processId": "8912",
      "resourceGuid": "resourceGuid",
      "resourceId": "/SUBSCRIPTIONS/subscriptionID goes here/RESOURCEGROUPS/resource group goes here/PROVIDERS/MICROSOFT.INSIGHTS/COMPONENTS/{function app name}",
      "sdkVersion": "azurefunctions: 3.0.15571.0",
      "success": true,
      "triggerReason": "This function was programmatically called via the host APIs.",
    },
    "eventType": "AzureAppRequest",
    "timestamp": 1619736074123,
  },
]
`;
